cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,336 @@
+import com.linkedin.kafka.cruisecontrol.model.Broker;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModelStats;
+import com.linkedin.kafka.cruisecontrol.model.Replica;
+import java.util.Collections;
+import com.linkedin.kafka.cruisecontrol.monitor.ModelCompletenessRequirements;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.ACCEPT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.REPLICA_REJECT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionGoal.ChangeType.ADD;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionGoal.ChangeType.REMOVE;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.GoalUtils.MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING;
+
+/**
+ * SOFT GOAL: Generate leadership movement proposals to ensure that the number of leader replicas on each broker is
+ * <ul>
+ * <li>Under: (the average number of leader replicas per broker) * (1 + leader replica count balance percentage)</li>
+ * <li>Above: (the average number of leader replicas per broker) * Math.max(0, 1 - leader replica count balance percentage)</li>
+ * </ul>
+ * Also see: {@link com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig#LEADER_REPLICA_COUNT_BALANCE_THRESHOLD_CONFIG}
+ * and {@link #balancePercentageWithMargin()}.
+ */
+public class LeaderReplicaDistributionGoal extends AbstractGoal {
+  private static final Logger LOG = LoggerFactory.getLogger(LeaderReplicaDistributionGoal.class);
+  private static final double BALANCE_MARGIN = 0.9;
+  private final Set<Integer> _brokerIdsAboveBalanceUpperLimit;
+  private final Set<Integer> _brokerIdsUnderBalanceLowerLimit;
+  private double _avgLeaderReplicasOnAliveBroker;
+  private int _balanceUpperLimit;
+  private int _balanceLowerLimit;
+
+  /**
+   * Constructor for Leader Replica Distribution Goal.
+   */
+  public LeaderReplicaDistributionGoal() {
+    _brokerIdsAboveBalanceUpperLimit = new HashSet<>();
+    _brokerIdsUnderBalanceLowerLimit = new HashSet<>();
+  }
+
+  public LeaderReplicaDistributionGoal(BalancingConstraint balancingConstraint) {
+    this();
+    _balancingConstraint = balancingConstraint;
+  }
+
+  /**
+   * To avoid churns, we add a balance margin to the user specified rebalance threshold. e.g. when user sets the
+   * threshold to be {@link BalancingConstraint#leaderReplicaBalancePercentage()}, we use
+   * ({@link BalancingConstraint#leaderReplicaBalancePercentage()}-1)*{@link #BALANCE_MARGIN} instead.
+   * @return the rebalance threshold with a margin.
+   */
+  private double balancePercentageWithMargin() {
+    return (_balancingConstraint.leaderReplicaBalancePercentage() - 1) * BALANCE_MARGIN;
```

Comment:
Why does this function omit `goalViolationDistributionThresholdMultiplier` -- i.e. isn't this a distribution goal?

Reply:
I think this is get missed in rebase.